src/components/common/table/index.tsx

```diff
@@ -28,3 +32,90 @@ export default function Table<T extends TableRowType>(props: TableProps<T>): JSX
     ...(props.showPagination !== false ? { renderPaginationText } : {}),
   });
 }
+
+/**
+ * Ordered columns component that preserves reordered columns using callback functions.
+ * This table fetches column names from preferences and sets them if available.
+ * Also saves new columns order upon reordering, into user preferences (for the org scope).
+ */
+
+export type OrderPreserveableTableProps = {
+  setColumns: (columns: TableColumnType[]) => void;
+};
+
+export function OrderPreserveableTable<T extends TableRowType>(
+  props: TableProps<T> & OrderPreserveableTableProps
+): JSX.Element {
+  const [initialized, setInitialized] = useState<boolean>(false);
+  const { setColumns, onReorderEnd, columns, id, ...tableProps } = props;
+  const { selectedOrganization, orgPreferences, reloadOrgPreferences } = useOrganization();
+
+  const getPreferenceName = useCallback(() => `${id}-columns`, [id]);
+
+  const getTableColumns = useCallback(
+    (columnNames: string[]): TableColumnType[] =>
+      columnNames.map((columnName) => columns.find((column) => column.key === columnName)) as TableColumnType[],
+    [columns]
+  );
+
+  const reorderHandler = async (reorderedColumns: string[]) => {
+    await PreferencesService.updateUserOrgPreferences(selectedOrganization.id, {
+      [getPreferenceName()]: reorderedColumns,
+    });
+    reloadOrgPreferences();
+    const columnsToSet = getTableColumns(reorderedColumns);
+    setColumns(columnsToSet);
+
+    if (onReorderEnd) {
+      onReorderEnd(reorderedColumns);
+    }
+  };
+
+  useEffect(() => {
+    const fetchSavedColumns = () => {
+      if (!orgPreferences) {
+        return;
+      }
+      setInitialized(true);
+      const columnNames: string[] | undefined = orgPreferences[getPreferenceName()] as string[] | undefined;
+      if (
+        columnNames?.length &&
+        columns.length === columnNames.length &&
+        !_.isEqual(
+          columns.map((column) => column.key),
+          columnNames
+        )
+      ) {
+        const columnsToSet = getTableColumns(columnNames);
+        setColumns(columnsToSet);
+      }
+    };
+
+    if (!initialized) {
+      fetchSavedColumns();
+    }
+  });
+
+  return BaseTable<T>({
+    ...tableProps,
+    id,
+    columns,
+    onReorderEnd: reorderHandler,
+  });
+}
+
+/**
+ * Ordered columns table with implementation to set columns - this handles the most common cases
+ * where client does not need extra semantics to filter columns.
+ * Species table is an example where this won't work, uses its own setColumns implementation.
+ */
+export default function OrderPreservedTable<T extends TableRowType>(props: TableProps<T>): JSX.Element {
```

Comment:
This is the new default table in terraware web. It uses the wrapped table that saves/initializes-from preferences.. with a default implementation to set the columns in state. Nothing fancy but doing it this way allows us to avoid changing any of the existing tables that conform to this generic use-case (no special semantics in setting columns order).